source/d3d9/buffer_detection.cpp

```diff
@@ -121,26 +129,48 @@ void reshade::d3d9::buffer_detection::on_clear_depthstencil(UINT clear_flags)
 	com_ptr<IDirect3DSurface9> depthstencil;
 	_device->GetDepthStencilSurface(&depthstencil);
 
-	if (depthstencil != _depthstencil_original && depthstencil != _depthstencil_replacement)
-		return; // Can only avoid clear of the replacement surface
-
-	auto &counters = _counters_per_used_depth_surface[_depthstencil_original];
+	depthstencil = (depthstencil == _depthstencil_replacement) ? _depthstencil_original : depthstencil;
 
-	// Ignore clears when there was no meaningful workload
-	// Also triggers when '_preserve_depth_buffers' is false, since no clear stats are recorded then
-	if (counters.current_stats.vertices <= 4 || counters.current_stats.drawcalls == 0)
+	if (depthstencil == nullptr || _depthstencil_replacement == nullptr || depthstencil != depthstencil_clear_index.first)
 		return;
 
-	counters.clears.push_back(counters.current_stats);
+	// always bound the stats to the original depthstencil surface
+	auto &counters = _counters_per_used_depth_surface[depthstencil];
 
-	if (depthstencil_clear_index == counters.clears.size())
+	// Update stats with data from previous frame
+	if (counters.current_stats.drawcalls == 0 && _first_empty_stats)
 	{
-		// Bind the original surface again so the clear is not performed on the replacement
-		_device->SetDepthStencilSurface(_depthstencil_original.get());
+		counters.current_stats.drawcalls = _previous_stats.drawcalls;
+		counters.current_stats.vertices = _previous_stats.vertices;
+		_first_empty_stats = false;
 	}
 
-	// Reset draw call stats for clears
-	counters.current_stats = { 0, 0 };
+	if (counters.current_stats.vertices <= 4 || counters.current_stats.drawcalls == 0) // Also triggers when '_preserve_depth_buffers' is false, since no clear stats are recorded then
+		return; // Ignore clears when there was no meaningful workload since the last one
+
+	counters.clears.push_back(counters.current_stats);
+
+	// switch to another depthsurface replacement surface
+	switch_depthsurface(depthstencil, counters);
```

Comment:
Considering that I don't think this has to be called in `buffer_detection::reset`, this would be the only other place to call `switch_depthsurface`, so may as well inline it here.

Reply:
Yes, but as i call it twice, i have prefered to put this on a specific method. And i found it more clear, because, that's exactly what it does, it switches to another dethsurface each time it is called.